intermediate_source/imbalanced_data_tutorial.py

```diff
@@ -0,0 +1,512 @@
+# -*- coding: utf-8 -*-
+"""
+Classifying an imbalanced dataset
+*********************************
+**Author**: `Piotr Bialecki <https://github.com/ptrblck>`_
+
+This tutorial show how to use the `WeightedRandomSampler` and a weighted loss
+function to classify an imbalanced dataset.
+
+We will load the CIFAR10 dataset and resample some classes creating an
+artificially imbalanced dataset. Using this dataset we will train a CNN and
+see how poorly the model performs on the undersampled class instances.
+
+In the next step we will create a `WeightedRandomSampler` and oversample the
+minority classes, so that our `DataLoader` will yield samples from nearly
+uniformly distributed classes.
+
+Finally, we will use a weighted criterion and train our model again on the
+imbalanced dataset.
+
+Have a look at the `CIFAR10 tutorial 
+<https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html>`_ if you
+would like to get more information about the dataset.
+
+Since we have to train the model for a few epochs, the usage of a GPU is 
+recommended.
+
+Let's start by importing some libs.
+"""
+from __future__ import print_function
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from torch.utils.data import Dataset, DataLoader
+
+from torchvision import datasets, transforms
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+###############################################################################
+# Loading the data and creating a model
+# =====================================
+#
+# In this tutorial we will use the CIFAR10 dataset and a standard CNN model.
+# Both can be loaded using ``torchvision``, but feel free to use your own
+# dataset or model.
+
+SEED = 2809
+torch.manual_seed(SEED)
+np.random.seed(SEED)
+if torch.cuda.is_available():
+    torch.cuda.manual_seed(SEED)
+batch_size = 64
+
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+
+transform = transforms.Compose([
+    transforms.ToTensor(),
+    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
+])
+
+# Training dataset
+train_dataset = datasets.CIFAR10(
+    root='.', train=True, download=True, transform=transform)
+train_loader = DataLoader(
+    dataset=train_dataset,
+    batch_size=batch_size,
+    shuffle=True,
+    num_workers=2,
+    pin_memory=torch.cuda.is_available())
+
+test_dataset = datasets.CIFAR10(
+    root='.', train=False, download=True, transform=transform)
+test_loader = DataLoader(
+    dataset=test_dataset,
+    batch_size=batch_size,
+    shuffle=False,
+    num_workers=2,
+    pin_memory=torch.cuda.is_available())
+
+class_names = ('plane', 'car', 'bird', 'cat', 'deer', 'dog', 'frog', 'horse',
+               'ship', 'truck')
+nb_classes = len(class_names)
+
+###############################################################################
+# Let's have a look at the class distribution in the datasets.
+
+
+def get_labels_and_class_counts(labels_list):
```

Comment:
Can you somehow move plotting code to one function as well? `get_labels_and_class_counts` is also used as class method, so this function is also confusing.

Reply:
function could be named visualize_class_imbalance with dataset args for example.

Reply:
I use the barplot for the original distribution and the imbalanced one.
So I created a functon named `plot_class_distributions`.

`get_labels_and_class_counts` is created as a function and as a class method. 
I'm not sure, if I can get rid of the class method, since I need the targets and counts from the imbalanced dataset. Since in the last review I created an own `Dataset` with indices for the imbalanced classes, I have to get these indices back. The CIFAR10 dataset stored in `ImbalancedCIFAR10` has still its original targets and therefore counts.